ironfish/src/assets/assetsVerifier.ts

```diff
@@ -0,0 +1,81 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+import { createRootLogger, Logger } from '../logger'
+import { ErrorUtils } from '../utils'
+import { SetIntervalToken } from '../utils'
+import { AssetsVerificationApi, VerifiedAssets } from './assetsVerificationApi'
+
+export type AssetVerification = {
+  status: 'verified' | 'unverified' | 'unknown'
+}
+
+export class AssetsVerifier {
+  private readonly REFRESH_INTERVAL = 6 * 60 * 60 * 1000 // 6 hours
+
+  private readonly logger: Logger
+  private readonly api: AssetsVerificationApi
+
+  private started: boolean
+  private refreshToken?: SetIntervalToken
+  private verifiedAssets?: VerifiedAssets
+
+  constructor(options?: { apiUrl?: string; logger?: Logger }) {
+    this.logger = options?.logger ?? createRootLogger()
+    this.api = new AssetsVerificationApi({ url: options?.apiUrl })
+    this.started = false
+  }
+
+  start(): void {
+    if (this.started) {
+      return
+    }
+
+    this.started = true
+    void this.refreshLoop()
+  }
+
+  stop(): void {
+    if (!this.started) {
+      return
+    }
+
+    this.started = false
+
+    if (this.refreshToken) {
+      clearTimeout(this.refreshToken)
+    }
+  }
+
+  private async refreshLoop(): Promise<void> {
+    await this.refresh()
+
+    this.refreshToken = setTimeout(() => {
+      void this.refreshLoop()
+    }, this.REFRESH_INTERVAL)
+  }
+
+  private async refresh(): Promise<void> {
+    try {
+      if (this.verifiedAssets) {
+        await this.api.refreshVerifiedAssets(this.verifiedAssets)
+      } else {
+        this.verifiedAssets = await this.api.getVerifiedAssets()
+      }
+    } catch (error) {
+      this.logger.error(
+        `Error while fetching verified assets: ${ErrorUtils.renderError(error)}`,
+      )
+    }
+  }
+
+  verify(assetId: Buffer | string): AssetVerification {
+    if (!this.started || !this.verifiedAssets) {
+      return { status: 'unknown' }
+    } else if (this.verifiedAssets.isVerified(assetId)) {
+      return { status: 'verified' }
+    } else {
+      return { status: 'unverified' }
```

Comment:
Are we going to have a warning/malicious tag as well? If somehow assets are identified as intentional scam assets, seems like this service would host that info too. Not sure if this is in scope though.

Reply:
This is something that we briefly discussed among me and some other people. There is no strong consensus, so it's not in scope yet, but it may become reality in the future. By the way, this is exactly the reason why this return value is an object and not a string, so that, in a hypothetical future, we could return something like:

```js
{ status: 'flagged',
  flagReason: 'this is used for scams' }
```